### PR TITLE
Felicia/dis 1424 stream run updates and block message while run is happening

### PR DESCRIPTION
**Pull request by feliciachang**

### CODE REVIEW

File: src/session-backend/index.ts [hunk 1]
```diff
@@ -242,14 +207,68 @@ io.on('connection', async (socket: Socket) => {
...
+  messageWithContext += 'This is the user request: '
+  messageWithContext += message
+  messageWithContext += ' '
+  messageWithContext += 'Here are the existing shapes in the whiteboard: '
+  messageWithContext += JSON.stringify(shapes)
+  messageWithContext +=
+    ' The y axis goes from negative (top) to positive (bottom). The x axis goes from negative (left) to positive (right).'
+
+  return messageWithContext
+}
```
Comment (rolyatmax) [inline src/session-backend/index.ts:274]: Here's an example of another way to write this using a template string (and with slightly different instructions):

```
function addContext(msg: string) {
  return `\
You are a helpful AI Assistant whose job is to help your users create and edit shapes on a canvas. \
The canvas is a 2D plane with an x and y axis. The y axis goes from negative (top) to positive (bottom). The x axis goes from negative (left) to positive (right).

These are the current shapes on the canvas: 

${JSON.stringify(shapes, null, 2)}

And this is the user's request:

${msg}`
}
```

File: src/session-backend/index.ts [hunk 2]
```diff
@@ -149,6 +126,7 @@ async function startRun(): Promise<void> {
           try {
             fn(functionArgs)
           } catch (err) {
```
Comment (rolyatmax) [inline src/session-backend/index.ts:128]: I wonder if after calling the function the assistant requested, maybe if we passed in what the shapes array looks like after the call - as the tool output (e.g. `fnStatus = 'shapes: ' + JSON.stringify(shapes)`). Maybe it'll get better at producing the right answer?

File: src/session-backend/index.ts (hunk 1 above)
Comment (feliciachang) [reply src/session-backend/index.ts:274]: what do the back slashes do?

File: src/session-backend/index.ts [hunk 3]
```diff
@@ -161,12 +139,15 @@ async function startRun(): Promise<void> {
           }
         })
 
-        // send the tool outputs to openai, which will restart a run if there were errors or complete the run if it was successful
-        runResult = await openai.beta.threads.runs.submitToolOutputs(thread.id, run.id, {
-          tool_outputs: toolOutputs,
-        })
+        if (activeRun) {
+          // send the tool outputs to openai, which will restart a run if there were errors or complete the run if it was successful
```
Comment (rolyatmax) [inline src/session-backend/index.ts:143]: Do we know that this is true? Might the assistant decide to continue the run if the tool output was technically successful but maybe doesn't end up producing the result it was looking for?

File: src/session-backend/index.ts (hunk 2 above)
Comment (feliciachang) [reply src/session-backend/index.ts:128]: mmm good point

File: src/session-backend/index.ts [hunk 4]
```diff
@@ -178,11 +159,15 @@ async function startRun(): Promise<void> {
 // Start a socket IO server that can be used to communicate between the client and server
 const io = new Server(8080, { cors: { origin: '*' } })
 
-// This shapes array will save the state of the whiteboard for a user session
+// shapes saves the state of the whiteboard for a user session
```
Comment (rolyatmax) [inline src/session-backend/index.ts:162]: maybe `holds` instead of `saves`?

File: src/session-backend/index.ts [hunk 5]
```diff
@@ -242,14 +207,68 @@ io.on('connection', async (socket: Socket) => {
...
 
 async function sleep(ms: number): Promise<void> {
-  return new Promise((resolve, reject) => {
+  return new Promise((resolve) => {
     setTimeout(resolve, ms)
   })
 }
+
+// functions we call in pollRun depending on the required action by openai
+const functions: Record<string, (toolOutput: any) => void> = {
```
Comment (rolyatmax) [inline src/session-backend/index.ts:227]: I kinda think it makes sense for the tools definition on line 17, which must match these functions, to be defined down here next to these. A general guideline for making code easier to read is: "code that changes together should stay together".

File: src/session-backend/index.ts (hunk 1 above)
Comment (rolyatmax) [reply src/session-backend/index.ts:274]: Those escape the next character, which in this case is the newline. In other words, it ignores the newline in the resulting string.

```js
const foo = `hello \
world`

// is just the string: "hello world" (no newline)
```

File: src/session-backend/index.ts (hunk 3 above)
Comment (feliciachang) [reply src/session-backend/index.ts:143]: just deleted the later part

File: src/session-backend/index.ts [hunk 6]
```diff
@@ -261,14 +266,15 @@ const functions: Record<string, (toolOutput: any) => void> = {
...
+
+  ${message}.
+
+  These are the current shapes on the canvas:
+
+  ${JSON.stringify(shapes, null, 2)}
 
-  return messageWithContext
+  You are a helpful AI Assistant whose job is to help your users create and edit shapes on a canvas based on their instructions. The canvas is a 2D plane with an x and y axis. The y axis goes from negative (top) to positive (bottom). The x axis goes from negative (left) to positive (right). [0, 0] is in the middle of the screen.
+  `
```
Comment (rolyatmax) [inline src/session-backend/index.ts:279]: Oh, just realized we are using this in the instructions we are passing to the Assistant and also adding it to every message. I feel like we could probably drop the repeated instructions and just send the user's request and the current shapes on the canvas.

File: src/components/Header.tsx [hunk 7]
```diff
@@ -1,17 +1,28 @@
...
-        <h1 className="flex gap-4 font-light tracking-widest items-center text-md uppercase drop-shadow-lg text-slate-500">
-          <JamsocketLogo />
-          Whiteboard
-        </h1>
-        {children}
+      <div className="mx-auto max-w-7xl px-6 pt-4">
+        <div className="flex items-center justify-between">
+          <h1 className="flex gap-4 font-light tracking-widest items-center text-md uppercase drop-shadow-lg text-slate-500">
+            <JamsocketLogo />
+            Whiteboard
```
Comment (rolyatmax) [inline src/components/Header.tsx:17]: We're changing this to something like "Canvas" or "Shared Canvas", right?

File: src/app/layout.tsx [hunk 8]
```diff
@@ -4,8 +4,8 @@ import { Inter } from 'next/font/google'
 const inter = Inter({ subsets: ['latin'] })
 
 export const metadata = {
-  title: 'Jamsocket NextJS Tutorial',
-  description: "Let's build a collaborative whiteboard with Jamsocket & NextJS!",
+  title: 'Jamsocket OpenAI Assistant Demo',
+  description: "Let's build a collaborative whiteboard with Jamsocket & OpenAI's Assistant API!",
```
Comment (rolyatmax) [inline src/app/layout.tsx:8]: "Let's build a shared canvas..." instead of "whiteboard"?

**Review by rolyatmax — APPROVED** [review]
A few tiny comments there re: the whiteboard naming + whether or not we need to pass in the repeated instructions to the AI with every message,

File: src/components/Header.tsx (hunk 7 above)
Comment (feliciachang) [reply src/components/Header.tsx:17]: i just talked to paul about this and i think we're keeping the title the same but clarifying that it's a shared canvas in other places

Comment (rolyatmax) [reply src/components/Header.tsx:17]: Ahh gotcha gotcha - sounds good to me!

**Pull request merged**